libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/rotate.hpp

```diff
@@ -122,26 +365,284 @@ namespace hpx { namespace parallel { inline namespace v1 {
+          HPX_CONCEPT_REQUIRES_(
+                hpx::is_execution_policy<ExPolicy>::value &&
+                hpx::traits::is_forward_iterator<FwdIter>::value &&
+                hpx::traits::is_sentinel_for<Sent, FwdIter>::value
+          )>
+        // clang-format on
+        friend typename parallel::util::detail::algorithm_result<ExPolicy,
+            subrange_t<FwdIter, Sent>>::type
+        tag_fallback_dispatch(hpx::ranges::rotate_t, ExPolicy&& policy,
+            FwdIter first, FwdIter middle, Sent last)
+        {
+            static_assert(hpx::traits::is_forward_iterator<FwdIter>::value,
+                "Requires at least forward iterator.");
+
+            typedef std::integral_constant<bool,
+                hpx::is_sequenced_execution_policy<ExPolicy>::value ||
+                    !hpx::traits::is_bidirectional_iterator<FwdIter>::value>
+                is_seq;
+
+            return hpx::parallel::util::get_subrange<FwdIter, Sent>(
+                hpx::parallel::v1::detail::rotate<
+                    parallel::util::in_out_result<FwdIter, Sent>>()
+                    .call2(std::forward<ExPolicy>(policy), is_seq(), first,
+                        middle, last));
+        }
+
+        // clang-format off
+        template <typename Rng,
+            HPX_CONCEPT_REQUIRES_(hpx::traits::is_range<Rng>::value)>
+        // clang-format on
+        friend subrange_t<typename hpx::traits::range_iterator<Rng>::type,
+            typename hpx::traits::range_iterator<Rng>::type>
+        tag_fallback_dispatch(hpx::ranges::rotate_t, Rng&& rng,
+            typename hpx::traits::range_iterator<Rng>::type middle)
+        {
+            return hpx::parallel::util::get_subrange<
+                typename hpx::traits::range_iterator<Rng>::type,
+                typename hpx::traits::range_sentinel<Rng>::type>(
+                hpx::parallel::v1::detail::rotate<parallel::util::in_out_result<
+                    typename hpx::traits::range_iterator<Rng>::type,
+                    typename hpx::traits::range_sentinel<Rng>::type>>()
+                    .call(hpx::execution::seq, hpx::util::begin(rng), middle,
+                        hpx::util::end(rng)));
+        }
+
+        // clang-format off
+        template <typename ExPolicy, typename Rng,
+            HPX_CONCEPT_REQUIRES_(
+                hpx::is_execution_policy<ExPolicy>::value &&
+                hpx::traits::is_range<Rng>::value)>
+        // clang-format on
+        friend typename parallel::util::detail::algorithm_result<ExPolicy,
+            subrange_t<typename hpx::traits::range_iterator<Rng>::type,
+                typename hpx::traits::range_sentinel<Rng>::type>>::type
+        tag_fallback_dispatch(hpx::ranges::rotate_t, ExPolicy&& policy,
+            Rng&& rng, typename hpx::traits::range_iterator<Rng>::type middle)
+        {
+            typedef hpx::is_sequenced_execution_policy<ExPolicy> is_seq;
```

Comment:
This should be typedef std::integral_constant<bool,
                hpx::is_sequenced_execution_policy<ExPolicy>::value ||
                    !hpx::traits::is_bidirectional_iterator<typename hpx::traits::range_iterator<Rng>::type>::value>
                is_seq;